homework2/hw1.py

```diff
@@ -0,0 +1,62 @@
+"""
+Given a file containing text. Complete using only default collections:
+    1) Find 10 longest words consisting from largest amount of unique symbols
+    2) Find rarest symbol for document
+    3) Count every punctuation char
+    4) Count every non ascii char
+    5) Find most common non ascii char for document
+"""
+
+
+import re
+import string
+from typing import List
+
+ACS = [chr(i) for i in range(127)]
+
+
+def get_longest_diverse_words(file_path: str) -> List[str]:
+    words = []
+    with open(file_path, "r+", encoding="unicode-escape") as file:
+        data = file.read()
+        r = re.compile(r"(\w+)")
+        words = r.findall(data)
+        words = sorted(words, key=len)
+        if len(words) > 10:
+            return words[-11:-1]
+        elif len(words) > 1:
+            return words[-len(words) : -1]
+        else:
+            return words[0]
+
+
+def get_rarest_char(file_path: str) -> str:
+    a = {}
+    with open(file_path, "r+", encoding="unicode-escape") as file:
+        file = file.read()
+        for i in file:
+            if i not in a.keys():
+                a[i] = file.count(i)
+    return min(a, key=a.get)
+
+
+def count_punctuation_chars(file_path: str) -> int:
+    with open(file_path, "r+", encoding="unicode-escape") as file:
+        file = file.read()
+        return len([i for i in file if i in string.punctuation])
+
+
+def count_non_ascii_chars(file_path: str) -> int:
+    with open(file_path, "r", encoding="unicode-escape") as file:
+        file = file.read()
+        return len([i for i in file if i not in ACS])
```

Comment:
There is an `ord` function. You can check if the char is in ascii or not much faster with it.